libflux/src/semantic/nodes.rs

```diff
@@ -0,0 +1,567 @@
+    pub loc: Location<'a>,
+
+    pub assignment: VariableAssignment<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ExpressionStatement<'a> {
+    pub loc: Location<'a>,
+
+    pub expression: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ReturnStatement<'a> {
+    pub loc: Location<'a>,
+
+    pub argument: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct VariableAssignment<'a> {
+    pub loc: Location<'a>,
+
+    pub id: Identifier<'a>,
+    pub init: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct MemberAssignment<'a> {
+    pub loc: Location<'a>,
+
+    pub member: MemberExpression<'a>,
+    pub init: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct StringExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub parts: Vec<StringExpressionPart<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum StringExpressionPart<'a> {
+    Text(TextPart<'a>),
+    Expr(InterpolatedPart<'a>),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct TextPart<'a> {
+    pub loc: Location<'a>,
+
+    pub value: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct InterpolatedPart<'a> {
+    pub loc: Location<'a>,
+
+    pub expression: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ArrayExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub elements: Vec<Expression<'a>>,
+}
+
+// FunctionExpression represents the definition of a function
+#[derive(Debug, PartialEq, Clone)]
+pub struct FunctionExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub params: Vec<FunctionParameter<'a>>,
+    pub body: FunctionBody<'a>,
+}
+
+impl<'a> FunctionExpression<'a> {
+    pub fn pipe(&self) -> Option<&FunctionParameter<'a>> {
+        for p in &self.params {
+            if p.is_pipe {
+                return Some(p);
+            }
+        }
+        None
+    }
+
+    pub fn defaults(&self) -> Vec<&FunctionParameter<'a>> {
+        let mut ds = Vec::new();
+        for p in &self.params {
+            match p.default {
+                Some(_) => ds.push(p),
+                None => (),
+            }
+        }
+        ds
+    }
+}
+
+// FunctionParameter represents a function parameter.
+#[derive(Debug, PartialEq, Clone)]
+pub struct FunctionParameter<'a> {
+    pub loc: Location<'a>,
+
+    pub is_pipe: bool,
+    pub key: Identifier<'a>,
+    pub default: Option<Expression<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct BinaryExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::OperatorKind,
+    pub left: Expression<'a>,
+    pub right: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct CallExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub callee: Expression<'a>,
+    pub arguments: ObjectExpression<'a>,
```

Comment:
I don't think this should be an object expression. This says that every function implicitly takes a single object as a parameter. But this isn't consistent with the ability of a function to specify default or pipe arguments. Nor is it consistent with extensible records which is what our objects are now.

Reply:
Mmmm, I guess the ability of a function to specify default or pipe arguments should be in `FunctionExpr`, and that's where it is.

`CallExpr`, instead, is only to specify how a function is called.

> Nor is it consistent with extensible records which is what our objects are now

For this, how should I change the representation then? In my last commit I opted for `Vec<Property>`.